pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+			// update plan status
+			notFound = false
+			planStatus := i.Status.PlanStatus[n1]
+			planStatus.Status = ExecutionPending
+			for i2, p := range v.Phases {
+				planStatus.Phases[i2].Status = ExecutionPending
+				for i3 := range p.Steps {
+					i.Status.PlanStatus[n1].Phases[i2].Steps[i3].Status = ExecutionPending
+				}
+			}
+
+			i.Status.PlanStatus[n1] = planStatus // we cannot modify item in map, we need to reassign here
+
+			// update activePlan and instance status
+			i.Status.AggregatedStatus.Status = ExecutionPending
+			i.Status.AggregatedStatus.ActivePlanName = planName
+
+			break
+		}
+	}
+	if notFound {
+		return &InstanceError{fmt.Errorf("asked to execute a plan %s but no such plan found in instance %s/%s", planName, i.Namespace, i.Name), kudo.String("PlanNotFound")}
+	}
 
-// PhaseState captures the state of the rollout.
-type PhaseState string
+	// TODO in the future when we again support manual plan execution, snapshot should be saved only non non-manually executed plans
+	err := i.SaveSnapshot()
+	if err != nil {
+		return err
+	}
 
-// PhaseStateInProgress actively deploying, but not yet healthy.
-const PhaseStateInProgress PhaseState = "IN_PROGRESS"
+	return nil
+}
 
-// PhaseStatePending Not ready to deploy because dependent phases/steps not healthy.
-const PhaseStatePending PhaseState = "PENDING"
+// isUpgradePlan returns true if this could be an upgrade plan - this is just an approximation because deploy plan can be used for both
+func isUpgradePlan(planName string) bool {
+	return planName == "deploy" || planName == "upgrade"
+}
 
-// PhaseStateComplete deployed and healthy.
-const PhaseStateComplete PhaseState = "COMPLETE"
+// UpdateInstanceStatus updates `Status.PlanStatus` and `Status.AggregatedStatus` property based on the given plan
+func (i *Instance) UpdateInstanceStatus(planStatus *PlanStatus) {
+	for k, v := range i.Status.PlanStatus {
+		if v.Name == planStatus.Name {
+			i.Status.PlanStatus[k] = *planStatus
+			i.Status.AggregatedStatus.Status = planStatus.Status
+			if planStatus.Status.IsTerminal() {
+				i.Status.AggregatedStatus.ActivePlanName = ""
+			}
+		}
+	}
+}
 
-// PhaseStateError there was an error deploying the application.
-const PhaseStateError PhaseState = "ERROR"
+const snapshotAnnotation = "kudo.dev/last-applied-instance-state"
 
-// PhaseStateSuspend Spec was triggered to stop this plan execution.
-const PhaseStateSuspend PhaseState = "SUSPEND"
+// SaveSnapshot stores the current spec of Instance into the snapshot annotation
+// this information is used when executing update/upgrade plans, this overrides any snapshot that existed before
+func (i *Instance) SaveSnapshot() error {
+	jsonBytes, err := json.Marshal(i.Spec)
+	if err != nil {
+		return err
+	}
+	if i.Annotations == nil {
+		i.Annotations = make(map[string]string)
+	}
+	i.Annotations[snapshotAnnotation] = string(jsonBytes)
+	return nil
+}
+
+func (i *Instance) getSnapshotedSpec() (*InstanceSpec, error) {
```

Comment:
this is named past tense... could we just call it snapshot? and can we leave the "get" off?
```suggestion
func (i *Instance) snapshotSpec() (*InstanceSpec, error) {
```